### ISSUE TITLE
SSH deploy keys

### ISSUE THREAD

**Issue by jpds** [nlewo/comin#37]
Would be great to have SSH deploy key support:

- https://docs.github.com/en/authentication/connecting-to-github-with-ssh/managing-deploy-keys#deploy-keys

GitLab and Forgejo support this too.

**Comment by slenderq** [comment]
I would find this very useful. 
 
Something simple like exposing  the `GIT_SSH_COMMAND` of the service would make this possible.

**Comment by Arbel-arad** [comment]
this already works if you add
```nix
    systemd = {
      services = {
        comin = {
          environment = {
            GIT_SSH_COMMAND = "${pkgs.openssh}/bin/ssh -i <path to key>";
          };
        };
      };
    };
```

**Comment by tcurdt** [comment]
@Arbel-arad I just tried this ... but I am getting

```
level=error msg="Pull from remote 'origin' failed: authentication required"
```

```
systemctl show -p Environment comin
Environment="GIT_SSH_COMMAND=/nix/store/i0nlz4mcyxzxd96x5dv0zcy23z6xkvzy-openssh-10.0p2/bin/ssh -i /run/secrets/comin"
```

```
ls -la /run/secrets/comin
-r-------- 1 root root 399 May 18 15:50 /run/secrets/comin
```

If this works for you - what is different for you?

**Comment by Arbel-arad** [comment]
@tcurdt most likely the systemd service is not seeing keys in `/run`
i am using host keys that are normal files and they work.
`Environment="GIT_SSH_COMMAND=/nix/store/0xh3cvypx00csnmz1b0xzy0vhsgapw3k-openssh-10.0p2/bin/ssh -i /etc/ssh/ssh_host_ed25519_key"`
 maybe try adding the secret path as a `BindReadOnlyPaths` directive in the service?

**Comment by tcurdt** [comment]
I had to walk away from the computer to realize: you better use the git url - not the https one 🙄

But unfortunately it's still not working for me:

```
May 18 13:51:22 utm-arm comin[113128]: time="2025-05-18T13:51:22Z" level=error msg="Pull from remote 'origin' failed: error creating SSH agent: \"SSH agent requested but SSH_AUTH_SOCK not-specified\""
```

Even after specifying it like this

```
  systemd = {
    services = {
      comin = {
        environment = {
          GIT_SSH_COMMAND = "${pkgs.openssh}/bin/ssh -i /run/secrets/comin -o IdentitiesOnly=yes -o ForwardAgent=no";
        };
      };
    };
  };

  services.comin = {
    enable = true;
    flakeSubdirectory = "nix";
    debug = true;
    remotes = [
      {
        name = "origin";
        url = "git@github.com:foo/bar.git";
        poller.period = 10;
      }
    ];
  };
```

When I run this manually it works:


```
GIT_SSH_COMMAND="ssh -i /run/secrets/comin -o IdentitiesOnly=yes -o ForwardAgent=no" git clone git@github.com:foo/bar.git
```

I tested with both, the secrets mount and a key in `/etc/ssh`. So that does not seem to make a difference. 
I somehow have the feeling that the `GIT_SSH_COMMAND` is not being passed on for me - for some reason.

Which version are you using, @Arbel-arad ? Also 0.6.0?

**Comment by tcurdt** [comment]
Seems like `comin` uses a [go git](github.com/go-git/go-git) implementation.
And it seems like that one is not aware of `GIT_SSH_COMMAND` 

https://github.com/search?q=repo%3Ago-git%2Fgo-git%20GIT_SSH_COMMAND&type=code
https://github.com/go-git/go-git/discussions/1207#discussioncomment-10973201

So I am confused how this is working for you? 🤔

**Comment by Arbel-arad** [comment]
> Seems like `comin` uses a [go git](github.com/go-git/go-git) implementation.
> And it seems like that one is not aware of `GIT_SSH_COMMAND` 
> 
> https://github.com/search?q=repo%3Ago-git%2Fgo-git%20GIT_SSH_COMMAND&type=code
> https://github.com/go-git/go-git/discussions/1207#discussioncomment-10973201
> 
> So I am confused how this is working for you? 🤔
> 

i see the issue now...
what i did was only using SSH for specific flake inputs that are secret, which works since that pulls through nix, but the cloning comin does ignores the SSH command

good catch but i'm not sure how to fix this...

**Comment by tcurdt** [comment]
> good catch but i'm not sure how to fix this...

It seems `go-git` does support ssh keys.
I suspect the easiest would be implement native `comin` support for it.

https://github.com/go-git/go-git/blob/main/_examples/clone/auth/ssh/private_key/main.go

Along the lines of:

```
    remotes = [
      {
        name = "origin";
        url = "git@github.com:foo/bar.git";
        auth.ssh_private_key_path = config.sops.secrets.comin.path;
        poller.period = 10;
      }
    ];
```

At least on the first glance that doesn't look too hard to add.

**Comment by tcurdt** [comment]
I guess @nlewo needs to weigh in whether that could be an acceptable contribution.

**Comment by tcurdt** [comment]
@nlewo did you test this? because I was trying this

https://github.com/nlewo/comin/issues/37#issuecomment-2889288954

...and it did not work for me because`go-git` does not support `GIT_SSH_COMMAND`.

**Comment by Arbel-arad** [comment]
I'm not entirely sure how go-git handles SSH at all, their example:
```go
r, err := git.PlainClone(directory, false, &git.CloneOptions{
	Auth:     publicKeys,
	URL:      url,
	Progress: os.Stdout,
})
```
only passes the public keys but no path to private? maybe i'm getting this wrong...
from issues on their repo looks like using ssh-agent is more stable in any case
https://github.com/go-git/go-git/issues/218

**Comment by nlewo** [comment]
> @nlewo did you test this

No, i didn't test it. I will revert the commit. Thank you for your tests and sorry to have missed your comment 😢

**Comment by Arbel-arad** [comment]
@nlewo It only works on flake inputs, but not the flake itself. i missed that part where different fetchers are used between them...
looks like the best way to fix this would be adding an option for `ssh-agent-auth` that would tell go-git to use the agent instead of tokens, and keep the `GIT_SSH_COMMAND` too for inputs.

**Comment by tcurdt** [comment]
> looks like the best way to fix this would be adding an option for ssh-agent-auth that would tell go-git to use the agent instead of tokens, and keep the GIT_SSH_COMMAND too for inputs.

TBH I am not sure a ssh-agent makes sense in a non-interactive scenario.

The `go-git` example I have linked has all the bits required to add a `auth.ssh_private_key_path` to `comin`.
Which should not be a big deal to add this if we decide to go for it.

Given that the github oci registry does not not support anything but classic pat tokens 😞 going for ssh keys became a little important for me. So less pressure to get this working. But I still think it should be added eventually.

**Comment by Arbel-arad** [comment]
@tcurdt the go-git issue i mentioned earlier is why i thought ssh-agent may be a better solution.
the fix on their side hasn't been merged yet (issue has been open for almost 5 years)
looks like the agent is just more stable when using go-git.
otherwise we could also use the `git` command directly which will just work with `GIT_SSH_COMMAND`

**Comment by camja014** [comment]
Hi all,

This is a feature I needed as my git hosting platform (Sourcehut) has limited support for PATs, and strongly prefers the use of SSH keys for authentication. So I went ahead took a pass at adding this functionality and opened a [PR](https://github.com/nlewo/comin/pull/166) for it. I have tested it in my setup, and it seems to be working.

I believe the issue referenced by @Arbel-arad in [this comment](https://github.com/nlewo/comin/issues/37#issuecomment-2952248431) relates to go-git's ability to automatically scan for, and use SSH keys stored in directories such as `~/.ssh` as OpenSSH does. Using an explicit path to an SSH private key seems to be fully supported by go-git.
